src/app.js

```diff
@@ -215,48 +217,134 @@ export default function (app) {
-        var i = createElementFrom(node)
-        parent.replaceChild(i, element)
-        element = i
-      }
-    } else if (node.tag) {
-      updateElementData(element, node.data, oldNode.data)
+      var n = createElementFrom(newNode)
+      parent.replaceChild(n, element)
+      element = n
+
+    } else if (oldNode.tag){
+      updateElementData(element, newNode.data, oldNode.data)
+      patchChildren(element, oldNode, newNode)
+    }
+
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var keyedChildren = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      if (!oldNode.children[i].data) continue
+      if (!oldNode.children[i].data.key) continue
+      keyedChildren[oldNode.children[i].data.key] = true
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
+      return (
+        !!node.data &&
+        !!node.data.key &&
+        !!keyedChildren[node.data.key]
+      )
+    }
+
+    var newIndex = 0
+    var oldIndex = 0
+    var indexOffset = 0
+    var cache = {}
+    var oldChild
+    var newChild
+
+    var patchChild = function () {
+      patch(element, element.childNodes[oldIndex + indexOffset], oldChild, newChild)
+      oldIndex++
+      newIndex++
+    }
+
+    var insertChild = function () {
+      patch(element, element.childNodes[oldIndex + indexOffset], undefined, newChild)
+      indexOffset++
+      newIndex++
+    }
+
+    var removeChild = function () {
+      batchRemove(element, element.childNodes[oldIndex + indexOffset], oldChild)
+      oldIndex++
+    }
+
+    while (
+      newNode.children[newIndex] !== undefined ||
+      oldNode.children[oldIndex] !== undefined
+    ) {
+      newChild = newNode.children[newIndex]
+      oldChild = oldNode.children[oldIndex]
+
+      if (
+        newChild !== undefined &&
+        isKeyed(newChild) &&
+        cache[newChild.data.key]
+      ) {
+        var c = cache[newChild.data.key]
+        element.insertBefore(c.element, element.childNodes[oldIndex + indexOffset])
+        patch(element, c.element, c.node, newChild)
+        delete cache[newChild.data.key]
+        newIndex++
 
-      for (var i = 0; i < len || i < oldLen; i++) {
-        patch(element, element.childNodes[i], oldNode.children[i], node.children[i])
+      } else if (oldChild === undefined) {
+        insertChild()
+
+      } else if (newChild === undefined) {
+        removeChild()
+
+      } else if (!isKeyed(oldChild) && !isKeyed(newChild)) {
+        patchChild()
+
+      } else if (!isKeyed(oldChild)) {
+        removeChild()
+
+      } else if (!isKeyed(newChild)) {
+        insertChild()
+
+      } else if (oldChild.data.key === newChild.data.key) {
+        patchChild()
+
+      } else {
+        cache[oldChild.data.key] = {
```

Comment:
Hmm... not sure I follow. I'm happy to try it, but could you give me a code-snippet to explain more. What I'm envisioning is not smaller so I'm sure I'm misunderstanding :)